engine/collection/ingest/engine.go

```diff
@@ -65,97 +72,147 @@ func New(
 		},
 	)
 
+	// FIFO queue for transactions
+	queue, err := fifoqueue.NewFifoQueue(
+		fifoqueue.WithCapacity(int(config.MaxMessageQueueSize)),
+		fifoqueue.WithLengthObserver(func(len int) {
+			mempoolMetrics.MempoolEntries(metrics.ResourceTransactionIngestQueue, uint(len))
+		}),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("could not create transaction message queue: %w", err)
+	}
+	pendingTransactions := &engine.FifoMessageStore{FifoQueue: queue}
+
+	// define how inbound messages are mapped to message queues
+	handler := engine.NewMessageHandler(
+		logger,
+		engine.NewNotifier(),
+		engine.Pattern{
+			Match: func(msg *engine.Message) bool {
+				_, ok := msg.Payload.(*flow.TransactionBody)
+				if ok {
+					engMetrics.MessageReceived(metrics.EngineCollectionIngest, metrics.MessageTransaction)
+				}
+				return ok
+			},
+			Store: pendingTransactions,
+		},
+	)
+
 	e := &Engine{
-		unit:                 engine.NewUnit(),
 		log:                  logger,
 		engMetrics:           engMetrics,
 		colMetrics:           colMetrics,
 		me:                   me,
 		state:                state,
+		pendingTransactions:  pendingTransactions,
+		messageHandler:       handler,
 		pools:                pools,
 		config:               config,
 		transactionValidator: transactionValidator,
 	}
 
+	e.ComponentManager = component.NewComponentManagerBuilder().
+		AddWorker(e.processQueuedTransactions).
+		Build()
+
 	conduit, err := net.Register(engine.PushTransactions, e)
 	if err != nil {
 		return nil, fmt.Errorf("could not register engine: %w", err)
 	}
-
 	e.conduit = conduit
 
 	return e, nil
 }
 
-// Ready returns a ready channel that is closed once the engine has fully
-// started.
-func (e *Engine) Ready() <-chan struct{} {
-	return e.unit.Ready()
-}
+// Process processes a transaction message from the network and enqueues the
+// message. Validation and ingestion is performed in the processQueuedTransactions
+// worker.
+func (e *Engine) Process(channel network.Channel, originID flow.Identifier, event interface{}) error {
+	select {
+	case <-e.ComponentManager.ShutdownSignal():
+		return component.ErrComponentShutdown
```

Comment:
do we actually handle this sentinel somewhere? How about we just return `nil` in such case?

Reply:
I would rather return an error than `nil`, since the message is not being processed. We shouldn't assume the caller knows the component is shut down - even if the error isn't handled, it at least gives visibility that the component is not in a healthy state (eg. in network layer logs, which is where this would currently surface).

Reply:
I can't see where it will be useful, I mean in future iterations based on [FLIP](https://github.com/onflow/flow/blob/alex/networking-api-flip/flips/20210201-consuming-messages-from-network-layer.md) we won't return any error to network layer since it doesn't know what to do with it.

Reply:
What about `ProcessTransaction` for another example? If another component calls `ProcessTransaction` and the component has been shut down, should we return an error or `nil`? I think we should return an error, for the same reasons mentioned in the last comment.

(Currently we don't handle this at all and just process the transaction normally, which I'll fix.)

Reply:
I would say in case of `ProcessTransaction` it makes sense to return error since it's called from RPC and it can be reported to user.
Based on FLIP, I am mainly concerned about `Process` which is called directly from network layer and as I said, network layer have no clue what to do with error so it's better to omit it.

Reply:
Updated in https://github.com/onflow/flow-go/pull/2035/commits/2b6cd4f0b2dc296b6fa7d943d9ef2e057b456cec